### PR TITLE
UI Automation in Windows Console: Make UIA the default for Windows 10 1803 and later

### LINKED ISSUE #513: Some output in command consoels spelled with speak typed characters enabled
**Reported by aleksey_s on 2009-12-17 16:39**
Sometimes, NVDA is too verbose in console windows.
STR:
- open cmd
- run python interpreter
- type: print "hi"
  Currently, NVDA says
  h
  i
  hi
  It starts reading by letter and only then reads the entire message. This behaviour also exists in other console apps.
  **Blocked by #5027**
  **Blocking #2977, #4127, #5328**

### LINKED ISSUE #673: Problems with command consoles and object navigation
**Reported by jteh on 2010-05-26 22:41**
When you move to a command console with object navigation, NVDA connects to this console, which allows this console to be reviewed. While this is useful, it presents three problems:
1. When you move to the console using object navigation, the console is connected, but it isn't disconnected when you move away.
   - This is because there is a becomeNavigatorObject event, but no event for the reverse.
2. When object navigation causes NVDA to connect to a console, new text is reported for that console. New text should only be reported for focused consoles.
3. Because only one console can be connected at a time, if a console is focused and the user moves to another console with object navigation, the user can no longer access the focused console. That is, new text is not read nor is caret navigation, backspace, etc.

While issues 1 and 2 can be fixed, it is probably not possible to solve issue 3, as this is due to a limitation of the implementation in Windows. Therefore, the only option may be to not connect to consoles for review.

Work around: To restore functionality to a focused console, move the navigator object back to the focused console.

This functionality was introduced in 4808ab654c473343bac0a49b185128d8e31ce116.

### PR DESCRIPTION
**Pull request by codeofdusk**
### Link to issue number:
Builds on #9614. Closes #673 and #6291.

### Summary of the issue:
NVDA's UIA console support significantly improves performance and stability, so we should enable it by default.

### Description of how this pull request fixes the issue:
While UIA in consoles was added in Windows 10 1709, the initial implementation left much to be desired, particularly concerning caret movement and expansion to character.

A new `UIAUtils.shouldUseUIAConsole` function has been added which determines whether to use UIA in consoles:
* If the user has explicitly chosen to use UIA or legacy console support, the function returns `True` or `False` respectively.
* Otherwise, if the user is running at least Windows 10 1803, the function returns `True`, or `False` if not.

The "use UI Automation in the Windows Console when available" GUI setting has been changed to expose all options in the config spec (i.e. to automatically detect UIA/legacy or override this autodetection).

### Testing performed:
Tested the new GUI setting on Windows 10 1903.

### Known issues with pull request:
None.

### Change log entry:
== New Features ==
- Vastly improved performance and stability in Command Prompt, PowerShell, and the Windows Subsystem for Linux on Windows 10 version 1803 and later. (#9771)
- In Command Prompt, PowerShell, and the Windows Subsystem for Linux on Windows 10 version 1803 and later, the reporting of typed text that does not appear onscreen (such as passwords) can now be toggled in NVDA's advanced preferences. (#9649)

### CODE REVIEW

**Comment by codeofdusk** [comment]
CC @feerrenrut @leonardder @michaelDCurran

**Comment by josephsl** [comment]
Hi, if you can, can you test this in 20H1 builds too? I myself will do some testing in 20H1 via Windows Sandbox. Thanks.

File: source/UIAUtils.py
Comment (leonardder) [inline source/UIAUtils.py]: Please wrap this in tripplets

```suggestion
	"""Determines whether to use UIA in the Windows Console."""
```

File: source/UIAUtils.py
Comment (leonardder) [inline source/UIAUtils.py]: I think I prefer this to be False for windows 10 1903 and above, until we can make sure that only 1903 will be affected by this.

File: source/_UIAHandler.py
Comment (leonardder) [inline source/_UIAHandler.py]: I think you can remove these comments here and move this information to the doc string of `shouldUseUIAConsole`.

File: source/gui/settingsDialogs.py
Comment (leonardder) [inline source/gui/settingsDialogs.py]: Typo: windoes

File: source/gui/settingsDialogs.py
Comment (leonardder) [inline source/gui/settingsDialogs.py]: Please make this a list. Rather than using dict[key], you can use list.index(key)

File: source/gui/settingsDialogs.py
Comment (leonardder) [inline source/gui/settingsDialogs.py]: Isn't `auto` supposed to be the default here?

File: source/gui/settingsDialogs.py
Comment (leonardder) [inline source/gui/settingsDialogs.py]: Please remove this in favour of the list as suggested above. You can save that list as an attribute on the panel. There are numerous examples within the NVDA code base how this is handled, e.g. focus context presentation.

File: user_docs/en/userGuide.t2t
Comment (leonardder) [inline user_docs/en/userGuide.t2t]: Please start every new sentence on a new line.

File: source/gui/settingsDialogs.py
Comment (leonardder) [inline source/gui/settingsDialogs.py]: Could you make it so that this checkbox is hidden when the former combo box is set to legacy? See for an example how the open scratchpad dir button is activated when the checkbox is ticked.

**Review by LeonarddeR — DISMISSED** [review]
Good to see that this is getting in to shape.

File: source/UIAUtils.py
Comment (codeofdusk) [reply source/UIAUtils.py]: @michaelDCurran How confident are you that this will be fixed in 2003? We're at least 8.5 months from release...

File: source/gui/settingsDialogs.py
Comment (codeofdusk) [reply source/gui/settingsDialogs.py]: Doesn't this just get it from the config (which defaults to auto in any case)?

**Comment by codeofdusk** [comment]
Once #9773 is merged, I think this PR is ready for master (pending code review, of course).

**Comment by lukaszgo1** [comment]
It might be worth to have some discussion about when issue is considered closed. In my view one of the following should happen:
1. The issue is fixed - doesn't apply to any of the bugs listed above.
2. The issue is impossible to fix - also not applicable because whine proper fixes aren't easy in these cases it is doable.
3. The product which is affected by the issue is either EOL or NVDA developers decided to no longer support it. It is important to point out that both Windows 7 and Windows 8.1 would be supported by Microsoft until ad least 2023, and LTSB version  of Windows 10 1607 on which Server 2016 is based until 2026.

In short I am against closing issues with the previous console implementations until they are properly fixed. The same has been done when UIA for Word was introduced - no one closed any bug report for COM implementation.

File: source/gui/settingsDialogs.py
Comment (leonardder) [reply source/gui/settingsDialogs.py]: Looks like this comment ended up at the wrong place.

**Comment by feerrenrut** [comment]
@lukaszgo1 Could you be specific about which issue you don't think should be closed to make sure we are on the same page?

File: source/gui/settingsDialogs.py
Comment (feerrenrut) [inline source/gui/settingsDialogs.py]: To prevent confusion when/if we become unable to support legacy in a future windows build, I think it might be more user friendly to name this "Prefer lagacy"

File: user_docs/en/userGuide.t2t
Comment (feerrenrut) [inline user_docs/en/userGuide.t2t]: It might be worth explicitly differentiating from "Windows terminal (currently in preview)". Since Windows Terminal runs cmd, powershell and WSL, users could easily get confused here.

File: user_docs/en/userGuide.t2t
Comment (feerrenrut) [inline user_docs/en/userGuide.t2t]: I would change this to "or work in trusted environments" rather than in secure environments.

**Comment by lukaszgo1** [comment]
@feerrenrut well I meant all tickets listed in the PR description by @codeofdusk namely: #513, #673, #1682 and #6291.

File: source/UIAUtils.py
Comment (feerrenrut) [inline source/UIAUtils.py]: The docs for `isWin10` say `@note: this function will always return False for source copies of NVDA due to a Python bug.`

Is this true? if not, please update the docs for the `isWin10` function.

Comment (leonardder) [reply source/UIAUtils.py]: It is True for python 2 and false for python 3.

Comment (codeofdusk) [reply source/UIAUtils.py]: Yes, this is definitely at least true for Python 2.

Comment (feerrenrut) [reply source/UIAUtils.py]: Ah, I think we are all on the same page, but I just realised my question "is it true" is quite open to interpretation. I was asking if the docs are correct. Please correct me if I'm mistaken with the following:
- Python 2: running from source causes the `isWin10` function to behave incorrectly, it always returns False.
- Python 3: the bug is fixed, `isWin10` function returns True when running from source on a Windows 10 machine.

If this makes it into Python 2 code (which is currently not the plan), we would need to look at this carefully to see if there is any way we can improve the situation, given how prevalent using the console is for people working from source copies. For now, no need to do anything, Bill could you add this to your checklist for the end of your project?

File: user_docs/en/userGuide.t2t
Comment (feerrenrut) [inline user_docs/en/userGuide.t2t]: Please add this sentence on a new line.

File: user_docs/en/userGuide.t2t
Comment (feerrenrut) [inline user_docs/en/userGuide.t2t]: Great!

**Review by feerrenrut — APPROVED** [review]
Just a minor change to put a new sentence on a new line in the user guide, otherwise I'm happy for this to be merged. To ensure that the Python 3 merge goes smoothly, I'll hold off on merging this until we have merged back from threshold.

**Review by feerrenrut — APPROVED** [review]
Thanks @codeofdusk

**Pull request merged**
